File: workspaces/ws_002_unitconverterpro/android/app/src/main/java/com/appfactory/unitconverterpro/model/ConversionEngine.kt
package com.appfactory.unitconverterpro.model

import java.text.DecimalFormat
import kotlin.math.abs

/** Pure, stateless conversion math. All factors come from [UnitData]. */
object ConversionEngine {

    /**
     * Converts [value] from one unit to another within the same [category].
     * Returns null if either unit is not in the category.
     */
    fun convert(value: Double, from: ConvUnit, to: ConvUnit, category: UnitCategory): Double? {
        if (category.units.none { it.id == from.id }) return null
        if (category.units.none { it.id == to.id }) return null

        if (category.isTemperature) {
            return convertTemperature(value, from.id, to.id)
        }
        // Linear conversion: value -> base -> target.
        val base = value * from.factor
        return base / to.factor
    }

    /** Temperature conversion via Celsius as the pivot. */
    private fun convertTemperature(value: Double, fromId: String, toId: String): Double {
        val celsius = when (fromId) {
            "temperature.celsius" -> value
            "temperature.fahrenheit" -> (value - 32.0) * 5.0 / 9.0
            "temperature.kelvin" -> value - 273.15
            else -> value
        }
        return when (toId) {
            "temperature.celsius" -> celsius
            "temperature.fahrenheit" -> celsius * 9.0 / 5.0 + 32.0
            "temperature.kelvin" -> celsius + 273.15
            else -> celsius
        }
    }

    /**
     * Formats a numeric result for display. Uses scientific notation for very
     * large or very small magnitudes, otherwise a trimmed grouped decimal.
     */
    fun format(value: Double): String {
        if (value.isNaN()) return "—"
        if (value.isInfinite()) return if (value > 0) "∞" else "-∞"
        if (value == 0.0) return "0"

        val magnitude = abs(value)
        return if (magnitude >= 1e12 || magnitude < 1e-6) {
            DecimalFormat("0.######E0").format(value)
        } else {
            DecimalFormat("#,##0.########").format(value)
        }
    }

    /**
     * Parses user text into a Double, tolerating grouping separators and a
     * comma used as a decimal separator.
     */
    fun parse(text: String): Double? {
        val trimmed = text.trim()
        if (trimmed.isEmpty()) return null
        val normalized = trimmed
            .replace(" ", "")
            .replace(",", "")
        normalized.toDoubleOrNull()?.let { return it }
        // Retry treating the last comma as a decimal point.
        return trimmed.replace(" ", "").replace(",", ".").toDoubleOrNull()
    }

    /** Plain (no grouping) string of a value, suitable for re-seeding an input field. */
    fun plainString(value: Double): String {
        if (!value.isFinite()) return "0"
        val s = DecimalFormat("0.########").format(value)
        return s
    }
}

File: workspaces/ws_002_unitconverterpro/ios/UnitConverterPro/Models/ConversionEngine.swift
import Foundation

/// Pure, stateless conversion math. All factors come from `UnitData`.
enum ConversionEngine {

    /// Converts `value` from one unit to another within the same category.
    /// Returns `nil` if the units belong to different categories.
    static func convert(_ value: Double,
                        from: ConvUnit,
                        to: ConvUnit,
                        category: UnitCategory) -> Double? {
        guard category.units.contains(where: { $0.id == from.id }),
              category.units.contains(where: { $0.id == to.id }) else { return nil }

        if category.isTemperature {
            return convertTemperature(value, from: from.id, to: to.id)
        }
        if category.id == "fuelEconomy" {
            return convertFuelEconomy(value, from: from.id, to: to.id)
        }
        // Linear conversion: value -> base -> target.
        let base = value * from.factor
        return base / to.factor
    }

    // MARK: - Temperature

    /// Temperature conversion via Celsius as the pivot.
    private static func convertTemperature(_ value: Double,
                                           from: String,
                                           to: String) -> Double {
        // Step 1: incoming value -> Celsius.
        let celsius: Double
        switch from {
        case "temperature.celsius":    celsius = value
        case "temperature.fahrenheit": celsius = (value - 32) * 5.0 / 9.0
        case "temperature.kelvin":     celsius = value - 273.15
        default:                       celsius = value
        }
        // Step 2: Celsius -> target.
        switch to {
        case "temperature.celsius":    return celsius
        case "temperature.fahrenheit": return celsius * 9.0 / 5.0 + 32
        case "temperature.kelvin":     return celsius + 273.15
        default:                       return celsius
        }
    }

    // MARK: - Fuel Economy

    /// Fuel economy. Most units are linear against km/L, but L/100km is the
    /// reciprocal. We only have linear units in `UnitData`, so this stays linear,
    /// but the dedicated path keeps the door open and guards against zero.
    private static func convertFuelEconomy(_ value: Double,
                                           from: String,
                                           to: String) -> Double {
        guard let fromUnit = UnitData.fuelEconomy.unit(withID: from),
              let toUnit = UnitData.fuelEconomy.unit(withID: to) else { return value }
        let base = value * fromUnit.factor
        return base / toUnit.factor
    }

    // MARK: - Formatting

    /// Formats a numeric result for display. Uses scientific notation for very
    /// large or very small magnitudes, otherwise a trimmed decimal string.
    static func format(_ value: Double) -> String {
        if value.isNaN { return "—" }
        if value.isInfinite { return value > 0 ? "∞" : "-∞" }
        if value == 0 { return "0" }

        let magnitude = abs(value)
        if magnitude != 0 && (magnitude >= 1e12 || magnitude < 1e-6) {
            let f = NumberFormatter()
            f.numberStyle = .scientific
            f.maximumFractionDigits = 6
            f.exponentSymbol = "e"
            f.locale = Locale.current
            return f.string(from: NSNumber(value: value)) ?? String(value)
        }

        let f = NumberFormatter()
        f.numberStyle = .decimal
        f.minimumFractionDigits = 0
        f.maximumFractionDigits = 8
        f.usesGroupingSeparator = true
        f.locale = Locale.current
        return f.string(from: NSNumber(value: value)) ?? String(value)
    }

    /// Parses user text into a Double, tolerating the current locale's
    /// grouping/decimal separators and plain ASCII input.
    static func parse(_ text: String) -> Double? {
        let trimmed = text.trimmingCharacters(in: .whitespaces)
        if trimmed.isEmpty { return nil }

        let f = NumberFormatter()
        f.numberStyle = .decimal
        f.locale = Locale.current
        if let n = f.number(from: trimmed) { return n.doubleValue }

        // Fallback: treat comma as decimal separator, strip spaces.
        let normalized = trimmed
            .replacingOccurrences(of: " ", with: "")
            .replacingOccurrences(of: ",", with: ".")
        return Double(normalized)
    }
}
